pkg/canary/knative_controller.go

```diff
@@ -0,0 +1,161 @@
+package canary
+
+import (
+	"context"
+	"fmt"
+	flaggerv1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+	clientset "github.com/fluxcd/flagger/pkg/client/clientset/versioned"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+	knative "knative.dev/serving/pkg/client/clientset/versioned"
+)
+
+type KnativeController struct {
+	kubeClient    kubernetes.Interface
+	flaggerClient clientset.Interface
+	knativeClient knative.Interface
+}
+
+// IsPrimaryReady checks if the primary revision is ready
+func (kc *KnativeController) IsPrimaryReady(cd *flaggerv1.Canary) (bool, error) {
+	service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return true, fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	revisionName, exists := service.Annotations["flagger.app/primary-revision"]
+	if !exists {
+		return true, fmt.Errorf("service %s.%s primary revision annotation not found", cd.Spec.TargetRef.Name, cd.Namespace)
+	}
+
+	revision, err := kc.knativeClient.ServingV1().Revisions(cd.Namespace).Get(context.TODO(), revisionName, metav1.GetOptions{})
+	if err != nil {
+		return true, fmt.Errorf("revision %s.%s get query error: %w", revisionName, cd.Namespace, err)
+	}
+
+	if !revision.IsReady() {
+		return true, fmt.Errorf("revision %s.%s is not ready", revision.Name, cd.Namespace)
+	}
+
+	return true, nil
+}
+
+// IsCanaryReady checks if the canary revision is ready
+func (kc *KnativeController) IsCanaryReady(cd *flaggerv1.Canary) (bool, error) {
+	service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return true, fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	revision, err := kc.knativeClient.ServingV1().Revisions(cd.Namespace).Get(context.TODO(), service.Status.LatestCreatedRevisionName, metav1.GetOptions{})
```

Comment:
(not a flagger expert)

I think you'd want to guard against `LatestCreatedRevisionName` being equal to the value of `annotations[flagger.app/primary-revision]` - this would ensure you wait for a new revision to be created which becomes your canary.

Otherwise I'm thinking if there was ever a delay in the knative reconcilers it'll look at the wrong revision here

Reply:
Note: revisions have a `configurationGeneration` label